File: vital-configs/src/main/java/me/vitalframework/configs/VitalConfigLocation.java
package me.vitalframework.configs;

import lombok.NonNull;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.WorldCreator;

/**
 * Wrapper class to store location data to a config file.
 */
public class VitalConfigLocation {
    @VitalConfig.Property(String.class)
    public String world;

    @VitalConfig.Property(double.class)
    public double x;

    @VitalConfig.Property(double.class)
    public double y;

    @VitalConfig.Property(double.class)
    public double z;

    @VitalConfig.Property(float.class)
    public float pitch;

    @VitalConfig.Property(float.class)
    public float yaw;

    @NonNull
    public static VitalConfigLocation of(@NonNull Location location) {
        final var vitalConfigLocation = new VitalConfigLocation();

        vitalConfigLocation.world = location.getWorld().getName();
        vitalConfigLocation.x = location.getX();
        vitalConfigLocation.y = location.getY();
        vitalConfigLocation.z = location.getZ();
        vitalConfigLocation.pitch = location.getPitch();
        vitalConfigLocation.yaw = location.getYaw();

        return vitalConfigLocation;
    }

    @NonNull
    public Location toLocation() {
        var world = Bukkit.getWorld(this.world);

        if (world == null) {
            // load world if null
            world = new WorldCreator(this.world)
                    .createWorld();
        }

        return new Location(world, x, y, z, yaw, pitch);

    }
}

File: vital-configs/src/main/java/me/vitalframework/configs/VitalConfigLocation.kt
package me.vitalframework.configs

import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.WorldCreator

/**
 * Wrapper class to store location data to a config file.
 */
class VitalConfigLocation {
    companion object {
        fun of(location: Location): VitalConfigLocation {
            val vitalConfigLocation = VitalConfigLocation()

            vitalConfigLocation.world = location.world!!.name
            vitalConfigLocation.x = location.x
            vitalConfigLocation.y = location.y
            vitalConfigLocation.z = location.z
            vitalConfigLocation.pitch = location.pitch
            vitalConfigLocation.yaw = location.yaw

            return vitalConfigLocation
        }
    }

    @VitalConfig.Property(String::class)
    var world: String? = null

    @VitalConfig.Property(Double::class)
    var x: Double = 0.0

    @VitalConfig.Property(Double::class)
    var y: Double = 0.0

    @VitalConfig.Property(Double::class)
    var z: Double = 0.0

    @VitalConfig.Property(Float::class)
    var pitch: Float = 0f

    @VitalConfig.Property(Float::class)
    var yaw: Float = 0f

    fun toLocation(): Location {
        var world = Bukkit.getWorld(this.world!!)

        if (world == null) {
            // load world if null
            world = WorldCreator(this.world!!)
                .createWorld()
        }

        return Location(world, x, y, z, yaw, pitch)
    }
}